File: Topics/dp/knapsack_recursive.cpp
#include<iostream>
using namespace std;

int knapsack(int value[],int weight[],int Weight,int n){

    if(n==0||Weight==0)
    return 0;

    if(weight[n-1]>Weight)
    return knapsack(value,weight,Weight,n-1);

    else
        return max(value[n-1]+knapsack(value,weight,Weight-weight[n-1],n-1),
            knapsack(value,weight,Weight,n-1));

}

int main(){

    int value[]={60,100,120};
    int weight[]={10,20,30};
    int Weight=50;

    int n=3;

    int ans=knapsack(value,weight,Weight,n);

    cout<<ans<<endl;
}

File: 2/dp/BoundedKnapsack/StandardKnapsack/KnapsackRecursive.java
public class KnapsackRecursive{

    public static void main(String[] args) {
      
        int N = 3;
        int W = 4;
        int val[] = {1,2,3};
        int wt[] = {4,5,1};

        System.out.println("Expected is 3, actual is :"+ knapsack(W, wt, val, N));

    }

    public static int knapsack(int W, int[] wt, int[] val, int N){
                 
        if(N==0 || W==0){
            return 0;
        }

        if(wt[N-1]<=W){
            return Integer.max(val[N-1]+knapsack(W-wt[N-1], wt, val, N-1),
                                knapsack(W, wt, val, N-1));
        }
        else 
            return knapsack(W, wt, val, N-1);

    }


}